base_layer/core/src/validation/transaction_validators.rs

```diff
@@ -0,0 +1,155 @@
+/// have passed
+pub struct FullTxValidator<B: BlockchainBackend> {
+    factories: Arc<CryptoFactories>,
+    db: BlockchainDatabase<B>,
+}
+
+impl<B: BlockchainBackend> FullTxValidator<B>
+where B: BlockchainBackend
+{
+    pub fn new(factories: Arc<CryptoFactories>, db: BlockchainDatabase<B>) -> Self {
+        Self { factories, db }
+    }
+}
+
+impl<B: BlockchainBackend> Validation<Transaction, B> for FullTxValidator<B> {
+    fn validate(&self, tx: &Transaction) -> Result<(), ValidationError> {
+        verify_tx(tx, &self.factories)?;
+        verify_inputs(tx, self.db.clone())?;
+        let height = self
+            .db
+            .get_height()
+            .map_err(|e| ValidationError::CustomError(e.to_string()))?
+            .unwrap_or(0);
+        verify_timelocks(tx, height)?;
+        Ok(())
+    }
+}
+
+/// This validator will only check that inputs exists in the backend.
+pub struct InputTxValidator<B: BlockchainBackend> {
+    db: BlockchainDatabase<B>,
+}
+
+impl<B: BlockchainBackend> InputTxValidator<B>
+where B: BlockchainBackend
+{
+    pub fn new(db: BlockchainDatabase<B>) -> Self {
+        Self { db }
+    }
+}
+
+impl<B: BlockchainBackend> Validation<Transaction, B> for InputTxValidator<B> {
+    fn validate(&self, tx: &Transaction) -> Result<(), ValidationError> {
+        verify_inputs(tx, self.db.clone())?;
+        Ok(())
+    }
+}
+
+/// This validator will only check timelocks, it will check that kernel lock heights and output maturities have passed.
+pub struct TimeLockTxValidator<B: BlockchainBackend> {
+    db: BlockchainDatabase<B>,
+}
+
+impl<B: BlockchainBackend> TimeLockTxValidator<B>
+where B: BlockchainBackend
+{
+    pub fn new(db: BlockchainDatabase<B>) -> Self {
+        Self { db }
+    }
+}
+
+impl<B: BlockchainBackend> Validation<Transaction, B> for TimeLockTxValidator<B> {
+    fn validate(&self, tx: &Transaction) -> Result<(), ValidationError> {
+        let height = self
+            .db
+            .get_height()
+            .map_err(|e| ValidationError::CustomError(e.to_string()))?
+            .unwrap_or(0);
+        verify_timelocks(tx, height)?;
+        Ok(())
+    }
+}
+
+// This function verifies that the provided transaction is internally sound and that no funds were created in the
+// transaction.
+fn verify_tx(tx: &Transaction, factories: &CryptoFactories) -> Result<(), ValidationError> {
+    tx.validate_internal_consistency(factories, None)
+        .map_err(|e| ValidationError::TransactionError(TransactionValidationError::TransactionError(e)))
+}
+
+// This function checks that all the timelocks in the provided transaction pass. It checks kernel lock heights and
+// output maturities
+fn verify_timelocks(tx: &Transaction, current_height: u64) -> Result<(), ValidationError> {
+    if tx.max_restriction_height() > current_height + 1 {
+        return Err(ValidationError::TransactionError(
+            TransactionValidationError::MaturityError,
+        ));
+    }
```

Comment:
Output maturities are not checked as per the docstring

Reply:
Typo, should be input maturities

Reply:
`max_restriction_height ` is now called `min_spendable_height` and does check both.